src/metadata.cpp

```diff
@@ -112,6 +116,21 @@ bool CLSNAME::checkMetadata(const Metadata& data) const        \
 
 #include "metadata_constraints.cpp"
 
+std::string escapeNonPrintableChars(const std::string& s)
+{
+  std::ostringstream os;
+  os << std::hex;
+  for (const char c : s) {
+    if (std::isprint(c)) {
+      os << c;
+    } else {
+      const unsigned int charVal = static_cast<unsigned char>(c);
+      os << "\\x" << std::setw(2) << std::setfill('0') << charVal;
+    }
+  }
+  return os.str();
+}
+
```

Comment:
I think this can be moved in `tools.(cpp|h)`

Reply:
I was about to move it, but then noticed that it will result in some confusion due to the presence of the functions `urlEncode()` and `needsEscape()` in an unnamed namespace in `tools.cpp` - within that translation unit the verb `escape` will denote two different ways of escaping a char in a string.

Reply:
Ok. Let's move on on this. If we need the function elsewhere, we will think about its renaming (or not)